provisioning/device/src/transport_state_machine.ts

```diff
@@ -0,0 +1,282 @@
+            // nothing to do.
+            callback();
+          }
+        },
+        connecting: {
+          _onEnter: (callback) => {
+            debug('entering connecting state');
+            debug('calling transport _doConnectForFsm');
+            /* Codes_SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_002: [ `connect` shall call `_doConnectForFsm`. ] */
+            this._doConnectForFsm((err) => {
+              /* Codes_SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_003: [ If `_doConnectForFsm` fails, then `connect` shall fail. ] */
+              if (err) {
+                /* Codes_SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_011: [ `register` shall fail if the connection fails. ] */
+                this._fsm.transition('disconnecting', callback, err);
+              } else {
+                this._fsm.transition('connected', callback);
+              }
+            });
+          },
+          '*': () => this._fsm.deferUntilTransition()
+        },
+        connected: {
+          _onEnter: (callback, err, body, result) => {
+            debug('entering connected state');
+            callback(err, body, result);
+          },
+          connect: (callback) => {
+            /* Codes_SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_004: [ If the transport is already connected, then `connect` shall do nothing and call the `callback` immediately. ] */
+            // nothing to do.
+            callback();
+          },
+          disconnect: (callback) => {
+            this._fsm.transition('disconnecting', callback);
+          },
+          register: (callback, registrationId, authorization, requestBody, forceRegistration) => {
+            this._fsm.transition('sendingRegistrationRequest', callback, registrationId, authorization, requestBody, forceRegistration);
+          },
+        },
+        sendingRegistrationReqest: {
+          _onEnter: (callback, registrationId, authorization, requestBody, forceRegistration) => {
+            debug('entering sendingRegistrationRequest state');
+            this._registrationCallback = callback;
+            debug('calling transport _doFirstRegistrationRequestForFsm');
+            /* Codes_SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_012: [ `register` shall call `_doFirstRegistrationRequestForFsm`. ] */
+            this._doFirstRegistrationRequestForFsm(registrationId, authorization, requestBody, forceRegistration, (err, responseBody, result, pollingInterval) => {
+              if (this._registrationInProgress()) { // make sure we weren't cancelled before doing something with the response
+                if (err) {
+                  /* Codes_SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_013: [ If the registration request fails, `register` shall fail. ] */
+                  this._fsm.transition('connected', callback, err);
```

Comment:
>('connected', callback, err) [](start = 38, length = 28)

Not disconnect?

Reply:
guessing registration might fail without disconnecting the transport? this being said, the error might be a disconnection so maybe falling back to connected could be, in some situations, misleading?

---
In reply to: [146401341](https://github.com/Azure/azure-iot-sdk-node/pull/115#discussion_r146401341) [](ancestors = 146401341)

Reply:
That's my thinking, but I'm still noodling on which state to go to in case of error.  Some errors (like a 404) keep the connection, but others don't.  There may be a giant list of failures-that-cause-disconnect in my future. :(

---
In reply to: [146403801](https://github.com/Azure/azure-iot-sdk-node/pull/115#discussion_r146403801) [](ancestors = 146403801,146401341)